### ISSUE TITLE
DER-encoded CMS EnvelopedData structure using the CMSEnvelopedDataStreamGenerator 

### ISSUE THREAD

**Issue by defacto64** [bcgit/bc-java#1296]
Hello,

anybody knows if it's possible to generate a DER-encoded CMS EnvelopedData structure using the CMSEnvelopedDataStreamGenerator class?

**Comment by roy-basmacier** [comment]
No, CMSEnvelopedDataStreamGenerator produces a BER encodings. The total content length isn't known when the structure header is written, so a fixed length DER isn't possible from the streaming API. Try using the CMSEnvelopedDataGenerator instead and force DER when encoding:

```java
CMSEnvelopedData ed = edGen.generate(new CMSProcessableByteArray(data), encryptor);
byte[] der = ed.toASN1Structure().getEncoded(ASN1Encoding.DER);
```

**Comment by dghgit** [comment]
As of 1.85 the answer is yes (it will be in the next beta at https://downloads.bouncycastle.org/betas/ — at the time the earlier comment was written it was
  correct).

```
  CMSEnvelopedDataStreamGenerator now supports definite-length output when you can supply the content length up front:

  CMSEnvelopedDataStreamGenerator edGen = new CMSEnvelopedDataStreamGenerator();

  edGen.addRecipientInfoGenerator(new JceKeyTransRecipientInfoGenerator(recipientCert).setProvider("BC"));
  edGen.setEncoding(ASN1Encoding.DER);     // or ASN1Encoding.DL

// write exactly contentLength octets of plaintext
  OutputStream envOut = edGen.open(outputStream, contentLength,
      new JceCMSContentEncryptorBuilder(CMSAlgorithm.AES256_GCM).setProvider("BC").build());

  // write exactly contentLength octets of plaintext
  envOut.close();
```

  The length is used to pre-compute every enclosing header before any ciphertext flows, so the output is fully definite-length DER, produced while streaming —
  nothing is buffered, and content larger than a byte array can hold is fine. Some points to be aware of:

  - The plain-text content length must be known before you open the stream; writing more or fewer octets than declared fails with an IOException (at which point the output
  is unusable and must be discarded — the headers already committed to the declared length).
  - In definite-length mode the setBufferSize() and setBEREncodeRecipients() settings are ignored: an indefinite-length encoding may not appear inside a definite-length one (actually, I'll admit this is not strictly true, but we do avoid it, you could say it's a bit rude to do otherwise).
  - The no-length open() methods throw in definite-length mode, pointing you at the length-taking overloads.

  The same facility was added to CMSSignedDataStreamGenerator and CMSAuthEnvelopedDataStreamGenerator (issue #1482 has the details).

For content that comfortably fits in memory with an unknown length, we now recommend using the new setEncoding() method on CMSEnvelopedDataGenerator.
